[iOS] Refactor some code around `UIWKTextInteractionAssistant`

whsieh opened the pull request:
#### dea762339998f966d7cda547c1c4e78527fd2274
<pre>
[iOS] Refactor some code around `UIWKTextInteractionAssistant`
<a href="https://bugs.webkit.org/show_bug.cgi?id=262725">https://bugs.webkit.org/show_bug.cgi?id=262725</a>

Reviewed by Megan Gardner.

Refactor some SPI usage related to `UIWKTextInteractionAssistant`. See below for more details.

* Source/WTF/wtf/PlatformHave.h:

Remove `HAVE(TEXT_INTERACTION_WITH_CONTEXT_MENU_INTERACTION)`. This has been true for all platforms
where `USE(UICONTEXTMENU)` is also true, since iOS 16.

* Source/WebKit/Platform/spi/ios/TextInputSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove more SPI that&apos;s now unused.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:

Remove some unused typedefs, along with the `_contextMenuInteraction` ivar (see below).

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[UIGestureRecognizer _wk_isTapAndAHalf]):

Drop the `UI-` prefix when performing a class check for the tap-and-half gesture.

(-[WKContentView gestureRecognizer:canBePreventedByGestureRecognizer:]):

Greatly simplify this logic by removing logic to deal with the interaction assistant&apos;s
`-forcePressGesture`. This is `nil` for all iPad, iPhone (both 3D-touch-enabled and non-3D-touch
models), and Mac Catalyst, so there&apos;s no longer any need to consider this when checking for text
interaction gestures. Long pressing to select text now entirely relies on the variable loupe
gesture, whose name is `&quot;UITextInteractionNameInteractiveRefinement&quot;`.

(-[WKContentView gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:]):

Similarly, remove all uses of `-loupeGesture` and `-singleTapGesture`. Instead, cache these on the
content view as instance variables (based on gesture name) after adding them to the view under
`-setUpTextSelectionAssistant`.

(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):
(-[WKContentView setUpTextSelectionAssistant]):

Also log a release error message in the case where either text interaction gestures can no longer be
found by iterating over the content view&apos;s gestures; this makes it a bit easier to determine when
(and if) this logic needs to be updated.

(-[WKContentView _translateForWebView:]):

Drive-by fix: remove a `-respondsToSelector` check that&apos;s no longer necessary.

(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView changeSelectionWithTouchesFrom:to:withGesture:withState:]):
(logTextInteractionAssistantSelectionChange):
(-[WKContentView _selectPositionAtPoint:stayingWithinFocusedElement:completionHandler:]):
(-[WKContentView selectPositionAtBoundary:inDirection:fromPoint:completionHandler:]):
(-[WKContentView moveSelectionAtBoundary:inDirection:completionHandler:]):
(-[WKContentView selectTextWithGranularity:atPoint:completionHandler:]):
(-[WKContentView beginSelectionInDirection:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPoint:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPoint:withBoundary:completionHandler:]):
(-[WKContentView _updateChangedSelection:]):

Remove logic to call into `-tintColorDidChange` on the selection view. `-selectionView` is actually
unconditionally `nil` on iOS 17 now, so this is just a no-op. I verified by manual testing (and also
by running the layout test `editing/caret/ios/caret-color-after-refocusing-input.html`) that this
caret-color bug remains fixed.

(-[WKContentView shouldDeferGestureDueToImageAnalysis:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):
(-[WKContentView cancelActiveTextInteractionGestures]):
(-[WKContentView selectPositionAtPoint:withContextRequest:completionHandler:]):
(-[WKContentView contextMenuInteraction]):

Remove the `_contextMenuInteraction` ivar from the content view entirely. This is no longer
necessary because the context menu interaction in `WKWebView` is added indirectly by the text
interaction assistant, and is exposed via `-[UIWKTextInteractionAssistant contextMenuInteraction]`.

(-[WKContentView _registerPreview]):
(-[WKContentView _unregisterPreview]):
(-[WKContentView textInteractionAssistantContextMenuInteraction]): Deleted.

Canonical link: <a href="https://commits.webkit.org/268955@main">https://commits.webkit.org/268955@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7264474e16a5a58b961d8d7fec80841a7fa99a98

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21127 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21492 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22190 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23004 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19644 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21688 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21351 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21065 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18322 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23858 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18223 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19152 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25418 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/18398 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19304 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19354 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23353 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/20540 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19891 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16912 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/24536 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19172 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5808 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5067 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23459 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/25801 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19750 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5634 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -4690,7 +4694,7 @@ - (void)changeSelectionWithGestureAt:(CGPoint)point withGesture:(UIWKGestureType
 - (void)changeSelectionWithGestureAt:(CGPoint)point withGesture:(UIWKGestureType)gestureType withState:(UIGestureRecognizerState)state withFlags:(UIWKSelectionFlags)flags
 {
-    logTextInteractionAssistantSelectionChange(__PRETTY_FUNCTION__, _textInteractionAssistant.get(), point, state, gestureType, std::nullopt, flags);
+    logTextInteractionAssistantSelectionChange(__PRETTY_FUNCTION__, _textInteractionLoupeGestureRecognizer.get(), point, state, gestureType, std::nullopt, flags);
```
I would've said remove `__PRETTY_FUNCTION__` and just use `std::source_location` but alas... rdar://116228776

whsieh replied:
TIL! But alas :(

whsieh committed:
[iOS] Refactor some code around `UIWKTextInteractionAssistant`
https://bugs.webkit.org/show_bug.cgi?id=262725

Reviewed by Megan Gardner.

Refactor some SPI usage related to `UIWKTextInteractionAssistant`. See below for more details.

* Source/WTF/wtf/PlatformHave.h:

Remove `HAVE(TEXT_INTERACTION_WITH_CONTEXT_MENU_INTERACTION)`. This has been true for all platforms
where `USE(UICONTEXTMENU)` is also true, since iOS 16.

* Source/WebKit/Platform/spi/ios/TextInputSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove more SPI that's now unused.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:

Remove some unused typedefs, along with the `_contextMenuInteraction` ivar (see below).

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[UIGestureRecognizer _wk_isTapAndAHalf]):

Drop the `UI-` prefix when performing a class check for the tap-and-half gesture.

(-[WKContentView gestureRecognizer:canBePreventedByGestureRecognizer:]):

Greatly simplify this logic by removing logic to deal with the interaction assistant's
`-forcePressGesture`. This is `nil` for all iPad, iPhone (both 3D-touch-enabled and non-3D-touch
models), and Mac Catalyst, so there's no longer any need to consider this when checking for text
interaction gestures. Long pressing to select text now entirely relies on the variable loupe
gesture, whose name is `"UITextInteractionNameInteractiveRefinement"`.

(-[WKContentView gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:]):

Similarly, remove all uses of `-loupeGesture` and `-singleTapGesture`. Instead, cache these on the
content view as instance variables (based on gesture name) after adding them to the view under
`-setUpTextSelectionAssistant`.

(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):
(-[WKContentView setUpTextSelectionAssistant]):

Also log a release error message in the case where either text interaction gestures can no longer be
found by iterating over the content view's gestures; this makes it a bit easier to determine when
(and if) this logic needs to be updated.

(-[WKContentView _translateForWebView:]):

Drive-by fix: remove a `-respondsToSelector` check that's no longer necessary.

(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView changeSelectionWithTouchesFrom:to:withGesture:withState:]):
(logTextInteractionAssistantSelectionChange):
(-[WKContentView _selectPositionAtPoint:stayingWithinFocusedElement:completionHandler:]):
(-[WKContentView selectPositionAtBoundary:inDirection:fromPoint:completionHandler:]):
(-[WKContentView moveSelectionAtBoundary:inDirection:completionHandler:]):
(-[WKContentView selectTextWithGranularity:atPoint:completionHandler:]):
(-[WKContentView beginSelectionInDirection:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPoint:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPoint:withBoundary:completionHandler:]):
(-[WKContentView _updateChangedSelection:]):

Remove logic to call into `-tintColorDidChange` on the selection view. `-selectionView` is actually
unconditionally `nil` on iOS 17 now, so this is just a no-op. I verified by manual testing (and also
by running the layout test `editing/caret/ios/caret-color-after-refocusing-input.html`) that this
caret-color bug remains fixed.

(-[WKContentView shouldDeferGestureDueToImageAnalysis:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):
(-[WKContentView cancelActiveTextInteractionGestures]):
(-[WKContentView selectPositionAtPoint:withContextRequest:completionHandler:]):
(-[WKContentView contextMenuInteraction]):

Remove the `_contextMenuInteraction` ivar from the content view entirely. This is no longer
necessary because the context menu interaction in `WKWebView` is added indirectly by the text
interaction assistant, and is exposed via `-[UIWKTextInteractionAssistant contextMenuInteraction]`.

(-[WKContentView _registerPreview]):
(-[WKContentView _unregisterPreview]):
(-[WKContentView textInteractionAssistantContextMenuInteraction]): Deleted.

Canonical link: https://commits.webkit.org/268955@main
